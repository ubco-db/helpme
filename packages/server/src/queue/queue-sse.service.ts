import { Role, SSEQueueResponse } from '@koh/common';
import { Injectable, Inject, forwardRef } from '@nestjs/common';
import { Response } from 'express';
import { throttle } from 'lodash';
import { SSEService } from 'sse/sse.service';
import { QueueService } from './queue.service';
import { QueueChatService } from 'queueChats/queue-chats.service';
type QueueClientMetadata = { userId: number; role: Role };

const idToRoom = (queueId: number) => `q-${queueId}`;
/**
 * Handle sending queue sse events
 *
 * Exposes the following:
 * - subscribeClient: subscribes someone to a redis room for a particular queue (they will then be subscribed to a particular queueId).
 *   Then, when a 'pub' event to a particular queueId happens in redis, this will take the new data and pump it into res to be sent to the client over a websocket.
 * - updateQuestions/updateQueue/updateQueueChats: get the latest data for this queue and send it to all clients that are subscribed
 *  to the redis room
 */
@Injectable()
export class QueueSSEService {
  constructor(
    private queueService: QueueService,
    private sseService: SSEService<QueueClientMetadata>,
    @Inject(forwardRef(() => QueueChatService))
    private queueChatService: QueueChatService,
  ) {}

  subscribeClient(
    queueId: number,
    res: Response,
    metadata: QueueClientMetadata,
  ): void {
    this.sseService.subscribeClient(idToRoom(queueId), res, metadata);
  }

  // Send event with new questions, but no more than once a second
  updateQuestions = this.throttleUpdate(async (queueId) => {
    const queueQuestions = await this.queueService.getQuestions(queueId);
    if (queueQuestions) {
      await this.sendToRoom(queueId, async ({ role, userId }) => ({
        queueQuestions: await this.queueService.personalizeQuestions(
          queueId,
          queueQuestions,
          userId,
          role,
        ),
      }));
    }
  });

  updateQueue = this.throttleUpdate(async (queueId) => {
    const queue = await this.queueService.getQueue(queueId);
    if (queue) {
      await this.sendToRoom(queueId, async () => ({ queue }));
    }
  });

  updateQueueChats = this.throttleUpdate(async (queueId) => {
    await this.sendToRoom(queueId, async ({ role, userId }) => ({
      queueChats: await this.queueChatService.getMyChats(queueId, role, userId),
    }));
  });

  /* Sends data to all clients in the room */
  private async sendToRoom(
    queueId: number,
    data: (metadata: QueueClientMetadata) => Promise<SSEQueueResponse>,
  ) {
    await this.sseService.sendEvent(idToRoom(queueId), data);
  }

  /**
   * This will throttle the update function to once every 150ms.
   *
   * This is helpful since some endpoints may call updateQueue or updateQuestions multiple times in a row, and this
   * will make it so only 1 updateFunction call is actually made (which is important, since basically every time
   * one of the update functions are called, it will send data to all subscribed browsers, causing re-renders in their browsers.
   * And so sending data that's nearly the same to browsers multiple times causes unnecessary expensive re-renders).
   *
   * However, despite this,for now I will have leading: true as well. This will make it so the first call of the function
   * will be called immediately, and the rest will be throttled. For example, for 3 calls, one at t=50ms, one at t=80ms, and one at t=100ms,
   * the first call will be called at t=50ms, and another will be called at t=200ms (making leading: false will make only 1 call at t=200ms).
   * This might make it more snappy, but may cause more unnecessary re-renders or unintended buggyness since the endpoint might not be finished
   * by the first time the function is called.
   * Set leading to false possibly in the future may be a good idea.
   */
  private throttleUpdate(updateFunction: (queueId: number) => Promise<void>) {
    return throttle(
      async (queueId: number) => {
        try {
          await updateFunction(queueId);
        } catch (e) {}
      },
      150, // Don't make too high, since it can cause tests to fail as it will run updateFunction even after the endpoint is finished
      {
        leading: true,
        trailing: true,
      },
    );
  }
}
