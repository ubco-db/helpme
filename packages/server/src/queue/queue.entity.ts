import { Exclude } from 'class-transformer';
import { QuestionGroupModel } from '../question/question-group.entity';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  JoinTable,
  ManyToMany,
  ManyToOne,
  OneToMany,
  OneToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CourseModel } from '../course/course.entity';
import { UserModel } from '../profile/user.entity';
import { QuestionModel } from '../question/question.entity';
import { HttpException, HttpStatus } from '@nestjs/common';
import {
  ERROR_MESSAGES,
  OpenQuestionStatus,
  QueueConfig,
  QueueTypes,
  StatusInQueue,
} from '@koh/common';
import { QuestionTypeModel } from '../questionType/question-type.entity';
import { QueueInviteModel } from './queue-invite.entity';
import { QueueChatsModel } from '../queueChats/queue-chats.entity';

@Entity('queue_model')
export class QueueModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => CourseModel, (course) => course.queues)
  @JoinColumn({ name: 'courseId' })
  course: CourseModel;

  @Column({ nullable: true })
  @Exclude()
  courseId: number;

  @Column('text')
  room: string; // queue name

  @OneToMany((type) => QuestionModel, (qm) => qm.queue)
  @Exclude()
  questions: QuestionModel[];

  @OneToMany((type) => QuestionGroupModel, (qg) => qg.queue)
  @Exclude()
  groups: QuestionGroupModel[];

  @Column('text', { nullable: true })
  notes: string;

  @ManyToMany((type) => UserModel, (user) => user.queues)
  @JoinTable()
  staffList: UserModel[];

  @Column({ default: false })
  allowQuestions: boolean;

  @Column({ default: 'hybrid' })
  type: QueueTypes;

  @Column('text', { nullable: true })
  zoomLink: string;

  @Column({ default: false })
  isProfessorQueue: boolean;

  @Column({ default: false })
  isDisabled: boolean; // TODO: replace with @deletedAt column

  @OneToMany((type) => QuestionTypeModel, (qtm) => qtm.queue)
  questionTypes: QuestionTypeModel[];

  // I really didn't want to make this a column
  // but typeorm no longer allows you to have a property that is not a column on an entity class.
  // It is a pain in the arse to fix this too,
  // since getQueue first adds on the queueSize with addQueueSize and then returns a QueueModel,
  // which other service functions will take, modify, and then run .save on it.
  // Other endpoints will just return what getQueue returns.
  // Basically, just know that you SHOULD NOT expect queueSize to be up-to-date in the database,
  // and that it's literally only there so that we can have this regular class attribute here.
  @Column({ default: 0 })
  queueSize: number;

  async addQueueSize(): Promise<void> {
    this.queueSize = await QuestionModel.inQueueWithStatus(this.id, [
      ...StatusInQueue,
      OpenQuestionStatus.Helping,
      OpenQuestionStatus.Paused,
    ]).getCount();
  }

  @Column('json', { nullable: true })
  config: QueueConfig;

  @OneToOne((type) => QueueInviteModel, (queueInvite) => queueInvite.queue, {
    cascade: true,
  })
  @Exclude()
  queueInvite: QueueInviteModel;

  @OneToMany(() => QueueChatsModel, (queueChat) => queueChat.queue)
  @Exclude()
  chats: QueueChatsModel[];
}
