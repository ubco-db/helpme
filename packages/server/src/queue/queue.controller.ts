import {
  ClosedQuestionStatus,
  ERROR_MESSAGES,
  GetQueueResponse,
  LimboQuestionStatus,
  ListQuestionsResponse,
  OpenQuestionStatus,
  QueueConfig,
  Role,
  UpdateQueueParams,
  isCycleInTasks,
  setQueueConfigResponse,
  validateQueueConfigInput,
} from '@koh/common';
import {
  Body,
  ClassSerializerInterceptor,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  NotFoundException,
  Param,
  ParseIntPipe,
  Patch,
  Post,
  Res,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { Response } from 'express';
import { UserId } from 'decorators/user.decorator';
import { JwtAuthGuard } from '../guards/jwt-auth.guard';
import { Roles } from '../decorators/roles.decorator';
import { QueueCleanService } from './queue-clean/queue-clean.service';
import { QueueRole } from '../decorators/queue-role.decorator';
import { QueueRolesGuard } from '../guards/queue-role.guard';
import { QueueSSEService } from './queue-sse.service';
import { QueueModel } from './queue.entity';
import { QueueService } from './queue.service';
import { QuestionModel } from '../question/question.entity';
import { EmailVerifiedGuard } from 'guards/email-verified.guard';
import { RedisQueueService } from '../redisQueue/redis-queue.service';

@Controller('queues')
@UseGuards(JwtAuthGuard, QueueRolesGuard, EmailVerifiedGuard)
@UseInterceptors(ClassSerializerInterceptor)
export class QueueController {
  constructor(
    private queueSSEService: QueueSSEService,
    private queueCleanService: QueueCleanService,
    private queueService: QueueService, //note: this throws errors, be sure to catch them
    private redisQueueService: RedisQueueService,
  ) {}

  /*
  Gets all queue info.
  Note there is a method that is very similar to this in queue-invite.controller.ts
  */
  @Get(':queueId')
  @Roles(Role.TA, Role.PROFESSOR, Role.STUDENT)
  async getQueue(
    @Param('queueId', ParseIntPipe) queueId: number,
  ): Promise<GetQueueResponse> {
    try {
      return this.queueService.getQueue(queueId);
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.getQueue,
        HttpStatus.NOT_FOUND,
      );
    }
  }

  /*
  Gets all questions in a queue and personalizes them.
  Note there is a method that is very similar to this in queue-invite.controller.ts
  */
  @Get(':queueId/questions')
  @Roles(Role.TA, Role.PROFESSOR, Role.STUDENT)
  async getQuestions(
    @Param('queueId') queueId: number,
    @QueueRole() role: Role,
    @UserId() userId: number,
  ): Promise<ListQuestionsResponse> {
    try {
      const queueKeys = await this.redisQueueService.getKey(`q:${queueId}`);
      let queueQuestions: any;

      if (Object.keys(queueKeys).length === 0) {
        console.log('Fetching from database');

        queueQuestions = await this.queueService.getQuestions(queueId);
        if (queueQuestions)
          await this.redisQueueService.setQuestions(
            `q:${queueId}`,
            queueQuestions,
          );
      } else {
        console.log('Fetching from Redis');
        queueQuestions = queueKeys.questions;
      }

      return await this.queueService.personalizeQuestions(
        queueId,
        queueQuestions,
        userId,
        role,
      );
    } catch (err) {
      console.log(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.getQuestions,
        HttpStatus.NOT_FOUND,
      );
    }
  }

  @Patch(':queueId')
  @Roles(Role.TA, Role.PROFESSOR)
  async updateQueue(
    @Param('queueId', ParseIntPipe) queueId: number,
    @Body() body: UpdateQueueParams,
  ): Promise<QueueModel> {
    const queue = await this.queueService.getQueue(queueId);
    if (queue === undefined) {
      throw new NotFoundException();
    }
    queue.type = body.type;
    queue.notes = body.notes;
    queue.allowQuestions = body.allowQuestions;
    queue.zoomLink = body.zoomLink;
    try {
      await queue.save();
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.saveQueue,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
    return queue;
  }

  @Post(':queueId/clean')
  @Roles(Role.TA, Role.PROFESSOR)
  async cleanQueue(
    @Param('queueId', ParseIntPipe) queueId: number,
  ): Promise<void> {
    // Clean up queue if necessary
    try {
      setTimeout(async () => {
        await this.queueCleanService.cleanQueue(queueId, true);
        await this.redisQueueService.deleteKey(`q:${queueId}`);
        await this.queueSSEService.updateQueue(queueId);
      });
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.cleanQueue,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  /**
   * Endpoint to tell frontend when the queue changes
   * Note there is a similar method in queue-invite.controller.ts
   *  */
  @Get(':queueId/sse')
  sendEvent(
    @Param('queueId', ParseIntPipe) queueId: number,
    @QueueRole() role: Role,
    @UserId() userId: number,
    @Res() res: Response,
  ): void {
    res.set({
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'X-Accel-Buffering': 'no',
      Connection: 'keep-alive',
    });

    try {
      this.queueSSEService.subscribeClient(queueId, res, { role, userId });
    } catch (err) {
      console.error(err);
    }
  }

  @Delete(':queueId')
  @Roles(Role.TA, Role.PROFESSOR)
  async disableQueue(
    @Param('queueId', ParseIntPipe) queueId: number,
    @QueueRole() role: Role,
  ): Promise<void> {
    // disable a queue
    const queue = await this.queueService.getQueue(queueId);
    if (!queue) {
      throw new NotFoundException();
    }

    if (queue.isProfessorQueue && role === Role.TA) {
      throw new HttpException(
        ERROR_MESSAGES.queueController.cannotCloseQueue,
        HttpStatus.UNAUTHORIZED,
      );
    }

    queue.isDisabled = true;

    // clear staff list
    queue.staffList = [];

    const questions = await QuestionModel.inQueueWithStatus(queueId, [
      ...Object.values(OpenQuestionStatus),
      ...Object.values(LimboQuestionStatus),
    ]).getMany();

    questions.forEach((q: QuestionModel) => {
      q.status = ClosedQuestionStatus.Stale;
      q.closedAt = new Date();
    });

    try {
      // try to save queue (and stale questions!)
      await QuestionModel.save(questions);
      await queue.save();
      await this.redisQueueService.deleteKey(`q:${queueId}`);
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.saveQueue,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  /**
   * Sets the JSON config for a queue and then returns success + any question tags that were created/deleted/updated
   */
  // note: for whatever reason, anytime you `return` a res.send(), you get "cannot set headers after they are sent" even though that's what you're supposed to do
  @Patch(':queueId/config')
  @Roles(Role.TA, Role.PROFESSOR)
  async setConfig(
    @Param('queueId', ParseIntPipe) queueId: number,
    @Body() newConfig: QueueConfig,
    @Res() res: Response,
  ): Promise<Response<setQueueConfigResponse>> {
    // make sure queue config is valid
    const configError = validateQueueConfigInput(newConfig);
    if (configError) {
      res.status(HttpStatus.BAD_REQUEST).send({ message: configError });
      return;
    }

    // Make sure there are no cycles in the tasks of the config
    if (newConfig.tasks && isCycleInTasks(newConfig.tasks)) {
      res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: ERROR_MESSAGES.queueController.cycleInTasks });
      return;
    }

    try {
      const questionTypeMessages =
        await this.queueService.updateQueueConfigAndTags(queueId, newConfig);

      res.status(HttpStatus.OK).send({ questionTypeMessages });
      return;
    } catch (err) {
      console.error(err); // internal server error: figure out what went wrong
      res
        .status(HttpStatus.INTERNAL_SERVER_ERROR)
        .send({ message: ERROR_MESSAGES.queueController.saveQueue });
      return;
    }
  }
}
