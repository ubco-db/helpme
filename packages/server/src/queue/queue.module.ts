import { forwardRef, Module } from '@nestjs/common';
import { QueueController } from './queue.controller';
import { SSEModule } from 'sse/sse.module';
import { QueueService } from './queue.service';
import { QueueSSEService } from './queue-sse.service';
import { QueueSubscriber } from './queue.subscriber';
import { AlertsService } from '../alerts/alerts.service';
import { AlertsModule } from '../alerts/alerts.module';
import { RedisQueueService } from 'redisQueue/redis-queue.service';
import { ApplicationConfigService } from '../config/application_config.service';
import { QueueInviteController } from './queue-invite.controller';
import { QueueChatService } from 'queueChats/queue-chats.service';
import { QueueChatsModule } from 'queueChats/queue-chats.module';
import { QueueCleanService } from './queue-clean/queue-clean.service';
import { QuestionModule } from '../question/question.module';
import { QuestionService } from '../question/question.service';
import { ApplicationConfigModule } from '../config/application_config.module';
import { RedisQueueModule } from '../redisQueue/redis-queue.module';

@Module({
  controllers: [QueueController, QueueInviteController],
  providers: [
    QueueService,
    ApplicationConfigService,
    RedisQueueService,
    QueueSSEService,
    QueueSubscriber,
    AlertsService,
    QueueCleanService,
    QuestionService,
  ],
  exports: [
    QueueSSEService,
    QueueCleanService,
    AlertsService,
    QuestionService,
    QueueService,
    RedisQueueService,
  ],
  imports: [
    ApplicationConfigModule,
    RedisQueueModule,
    SSEModule,
    AlertsModule,
    forwardRef(() => QueueChatsModule),
    forwardRef(() => QuestionModule),
  ],
})
export class QueueModule {}
