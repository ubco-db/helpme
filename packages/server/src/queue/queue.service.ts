import {
  ListQuestionsResponse,
  OpenQuestionStatus,
  Question,
  Role,
  StatusInPriorityQueue,
  StatusInQueue,
  StatusSentToCreator,
} from '@koh/common';
import { Injectable, NotFoundException } from '@nestjs/common';
import { classToClass } from 'class-transformer';
import { pick } from 'lodash';
import { QuestionModel } from 'question/question.entity';
import { In } from 'typeorm';
import { QueueModel } from './queue.entity';
import { AlertsService } from '../alerts/alerts.service';
import { ApplicationConfigService } from 'config/application_config.service';

/**
 * Get data in service of the queue controller and SSE
 * WHY? To ensure data returned by endpoints is *exactly* equal to data sent by SSE
 */
@Injectable()
export class QueueService {
  constructor(
    private alertsService: AlertsService,
    private readonly appConfig: ApplicationConfigService,
  ) {}

  async getQueue(queueId: number): Promise<QueueModel> {
    const queue = await QueueModel.findOne(queueId, {
      relations: ['staffList'],
    });
    await queue.checkIsOpen();
    await queue.addQueueSize();

    return queue;
  }

  async getQuestions(queueId: number): Promise<ListQuestionsResponse> {
    // todo: Make a student and a TA version of this function, and switch which one to use in the controller
    // for now, just return the student response
    const queueSize = await QueueModel.count({
      where: { id: queueId },
    });
    // Check that the queue exists
    if (queueSize === 0) {
      throw new NotFoundException();
    }

    const questionsFromDb = await QuestionModel.inQueueWithStatus(
      queueId,
      [...StatusInPriorityQueue, ...StatusInQueue, OpenQuestionStatus.Helping],
      this.appConfig.get('max_questions_per_queue'),
    )
      .leftJoinAndSelect('question.questionTypes', 'questionTypes')
      .leftJoinAndSelect('question.creator', 'creator')
      .leftJoinAndSelect('question.taHelped', 'taHelped')
      .getMany();

    const unresolvedRephraseQuestionAlerts =
      await this.alertsService.getUnresolvedRephraseQuestionAlert(queueId);

    const questions = new ListQuestionsResponse();

    questions.queue = questionsFromDb.filter((question) =>
      StatusInQueue.includes(question.status as OpenQuestionStatus),
    );

    questions.questionsGettingHelp = questionsFromDb.filter(
      (question) =>
        question.status === OpenQuestionStatus.Helping && !question.groupId,
    );

    questions.priorityQueue = questionsFromDb.filter((question) =>
      StatusInPriorityQueue.includes(question.status as OpenQuestionStatus),
    );

    questions.groups = [];

    questions.unresolvedAlerts = unresolvedRephraseQuestionAlerts.map(
      (alert) => alert.payload,
    );

    questions.queue = questions.queue.map((question) => {
      const temp = pick(question, [
        'id',
        'queueId',
        'text',
        'creatorId',
        'taHelpedId',
        'createdAt',
        'firstHelpedAt',
        'helpedAt',
        'closedAt',
        'status',
        'location',
        'groupable',
        'groupId',
        'questionTypes',
        'taHelped',
      ]);

      Object.assign(temp, {
        creator: {
          name: question.creator.name,
          photoURL: question.creator.photoURL,
        },
      });

      return temp as Question;
    });
    return questions;
  }

  /** Hide sensitive data to other students */
  async personalizeQuestions(
    queueId: number,
    questions: ListQuestionsResponse,
    userId: number,
    role: Role,
  ): Promise<ListQuestionsResponse> {
    if (role === Role.STUDENT) {
      const newLQR = new ListQuestionsResponse();
      Object.assign(newLQR, questions);

      newLQR.queue = questions.queue.map((question) => {
        const creator =
          question.creator.id === userId
            ? question.creator
            : pick(question.creator, ['id']);
        // classToClass transformer will apply the @Excludes
        return classToClass<Question>(
          QuestionModel.create({ ...question, creator }),
        );
      });

      newLQR.questionsGettingHelp = questions.questionsGettingHelp.map(
        (question) => {
          const creator =
            question.creator.id === userId
              ? question.creator
              : pick(question.creator, ['id']);
          // classToClass transformer will apply the @Excludes
          return classToClass<Question>(
            QuestionModel.create({ ...question, creator }),
          );
        },
      );

      newLQR.yourQuestion = await QuestionModel.findOne({
        relations: ['creator', 'taHelped'],
        where: {
          creatorId: userId,
          queueId: queueId,
          status: In(StatusSentToCreator),
        },
      });
      newLQR.priorityQueue = [];

      if (newLQR.yourQuestion) {
        const temp = pick(newLQR.yourQuestion, [
          'closedAt',
          'queueId',
          'createdAt',
          'creatorId',
          'firstHelpedAt',
          'groupId',
          'groupable',
          'helpedAt',
          'id',
          'location',
          'questionTypes',
          'queueId',
          'status',
          'taHelpedId',
          'taHelped',
          'text',
        ]);

        newLQR.yourQuestion = Object.assign(temp, {
          creator: {
            name: newLQR.yourQuestion.creator.name,
            photoURL: newLQR.yourQuestion.creator.photoURL,
          },
        }) as Question;
      }
      return newLQR;
    }
    return questions;
  }
}
