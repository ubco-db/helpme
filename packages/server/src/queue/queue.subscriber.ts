import { QueueSSEService } from './queue-sse.service';
import {
  DataSource,
  EntitySubscriberInterface,
  EventSubscriber,
  UpdateEvent,
} from 'typeorm';
import { QueueModel } from './queue.entity';

@EventSubscriber()
export class QueueSubscriber implements EntitySubscriberInterface<QueueModel> {
  private queueSSEService: QueueSSEService;
  constructor(dataSource: DataSource, queueSSEService: QueueSSEService) {
    this.queueSSEService = queueSSEService;
    dataSource.subscribers.push(this);
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  listenTo() {
    return QueueModel;
  }

  async afterUpdate(event: UpdateEvent<QueueModel>): Promise<void> {
    if (event.entity) {
      // Send all listening clients an update
      await this.queueSSEService.updateQueue(event.entity.id);
    }
  }
}
