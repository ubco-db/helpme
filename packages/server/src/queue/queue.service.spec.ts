import { ListQuestionsResponse, QuestionStatusKeys, Role } from '@koh/common';
import { Test, TestingModule } from '@nestjs/testing';
import { mapValues, zip } from 'lodash';
import { QuestionModel } from 'question/question.entity';
import { DataSource } from 'typeorm';
import {
  initFactoriesFromService,
  QuestionFactory,
  QueueFactory,
  UserFactory,
} from '../../test/util/factories';
import { TestConfigModule, TestTypeOrmModule } from '../../test/util/testUtils';
import { QueueModel } from './queue.entity';
import { QueueService } from './queue.service';
import { AlertsService } from '../alerts/alerts.service';
import { ApplicationTestingConfigModule } from 'config/application_config.module';
import { FactoryModule } from 'factory/factory.module';
import { FactoryService } from 'factory/factory.service';

describe('QueueService', () => {
  let service: QueueService;
  let dataSource: DataSource;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        TestTypeOrmModule,
        TestConfigModule,
        FactoryModule,
        ApplicationTestingConfigModule,
      ],
      providers: [QueueService, AlertsService],
    }).compile();

    service = module.get<QueueService>(QueueService);
    dataSource = module.get<DataSource>(DataSource);

    // Grab FactoriesService from Nest
    const factories = module.get<FactoryService>(FactoryService);
    // Initialize the named exports to point to the actual factories
    initFactoriesFromService(factories);
  });

  afterAll(async () => {
    await dataSource.destroy();
  });

  beforeEach(async () => {
    await dataSource.synchronize(true);
  });

  // create 1 question for each status that exists, and put them all in a queue
  async function createQuestionsEveryStatus(
    queue: QueueModel,
  ): Promise<QuestionModel[]> {
    const allStatuses = Object.values(QuestionStatusKeys);
    const questions = await QuestionFactory.createList(allStatuses.length, {
      queue,
    });
    for (const [status, question] of zip(allStatuses, questions)) {
      question.status = status;
    }
    await QuestionModel.save(questions);
    return questions;
  }

  describe('getQuestions', () => {
    it('only returns questions in the given queue', async () => {
      const queue = await QueueFactory.create();
      await QuestionFactory.create();
      await QuestionFactory.create({ queue });
      expect((await service.getQuestions(queue.id)).questions.length).toEqual(
        1,
      );
    });

    it('filters questions by status appropriately', async () => {
      const queue = await QueueFactory.create();
      await createQuestionsEveryStatus(queue);

      const questionResponse = await service.getQuestions(queue.id);
      const statuses = mapValues(
        questionResponse,
        (questions: QuestionModel[]) => questions.map((q) => q.status),
      );
      expect(statuses).toEqual({
        priorityQueue: ['PriorityQueued'],
        questionsGettingHelp: ['Helping', 'Paused'],
        questions: ['Queued', 'Drafting', 'ReQueueing'],
        groups: [],
        unresolvedAlerts: [],
      });
    });

    it('sorts queue questions by createdat', async () => {
      const queue = await QueueFactory.create();
      const questionIds = [];
      for (let i = 0; i < 3; i++) {
        const question = await QuestionFactory.create({
          queue,
          createdAt: new Date(Date.now() + i * 1000),
        });
        questionIds.push(question.id);
      }

      expect(
        (await service.getQuestions(queue.id)).questions.map((q) => q.id),
      ).toEqual(questionIds);
    });

    it('sorts priority queue questions by createdat', async () => {
      const queue = await QueueFactory.create();
      const questionIds = [];
      for (let i = 0; i < 3; i++) {
        const question = await QuestionFactory.create({
          queue,
          status: 'PriorityQueued',
          createdAt: new Date(Date.now() + i * 1000),
        });
        questionIds.push(question.id);
      }

      expect(
        (await service.getQuestions(queue.id)).priorityQueue.map((q) => q.id),
      ).toEqual(questionIds);
    });
  });

  describe('personalizeQuestions', () => {
    let queue;
    beforeEach(async () => {
      queue = await QueueFactory.create();
    });

    const personalize = (
      lqr: ListQuestionsResponse,
      userId: number,
      role: Role,
    ) => service.personalizeQuestions(queue.id, lqr, userId, role);

    it('does nothing if not a student', async () => {
      const user = await UserFactory.create();
      await QuestionFactory.create({
        queue,
        createdAt: new Date('2020-11-02T12:00:00.000Z'),
      });
      const lqr = await service.getQuestions(queue.id);
      expect(
        await service.personalizeQuestions(queue.id, lqr, user.id, Role.TA),
      ).toEqual(lqr);
    });

    it('adds yourQuestions for students with a question in the queue', async () => {
      const user = await UserFactory.create();
      // Create a question but not in this queue
      await QuestionFactory.create({ creator: user });

      const blank: ListQuestionsResponse = {
        questions: [],
        priorityQueue: [],
        questionsGettingHelp: [],
        groups: [],
      };
      let lqr = await personalize(blank, user.id, Role.STUDENT);
      expect(lqr.yourQuestions).toEqual([]);

      // Create a question in this queue
      const question = await QuestionFactory.create({
        creator: user,
        queue,
      });
      lqr = await personalize(blank, user.id, Role.STUDENT);
      expect(lqr.yourQuestions[0].id).toEqual(question.id);
    });

    it('hides details of other students', async () => {
      const ours = await QuestionFactory.create({
        queue,
        createdAt: new Date('2020-11-02T12:00:00.000Z'),
        text: 'help us',
      });
      await QuestionFactory.create({
        queue,
        createdAt: new Date('2020-11-02T12:00:00.000Z'),
        text: 'help someone else',
      });

      const lqr = await personalize(
        await service.getQuestions(queue.id),
        ours.creatorId,
        Role.STUDENT,
      );
      expect(lqr).toMatchSnapshot();
    });
  });
});
