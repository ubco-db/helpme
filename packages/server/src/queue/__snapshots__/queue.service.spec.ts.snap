// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueueService personalizeQuestions hides details of other students 1`] = `
ListQuestionsResponse {
  "groups": [],
  "priorityQueue": [],
  "questionsGettingHelp": [],
  "queue": [
    QuestionModel {
      "closedAt": null,
      "createdAt": 2020-11-02T12:00:00.000Z,
      "creator": UserModel {},
      "groupable": true,
      "helpedAt": null,
      "id": 1,
      "location": null,
      "questionTypes": [
        QuestionTypeModel {
          "cid": 1,
          "color": "#000000",
          "id": 1,
          "name": "Question Type",
          "queueId": 2,
        },
      ],
      "status": "Queued",
      "taHelped": null,
      "text": "help us",
    },
    QuestionModel {
      "closedAt": null,
      "createdAt": 2020-11-02T12:00:00.000Z,
      "creator": UserModel {},
      "groupable": true,
      "helpedAt": null,
      "id": 2,
      "location": null,
      "questionTypes": [
        QuestionTypeModel {
          "cid": 1,
          "color": "#000000",
          "id": 2,
          "name": "Question Type",
          "queueId": 3,
        },
      ],
      "status": "Queued",
      "taHelped": null,
      "text": "help someone else",
    },
  ],
  "unresolvedAlerts": [],
  "yourQuestion": {
    "closedAt": null,
    "createdAt": 2020-11-02T12:00:00.000Z,
    "creator": {
      "name": "User Person",
      "photoURL": null,
    },
    "creatorId": 1,
    "firstHelpedAt": null,
    "groupId": null,
    "groupable": true,
    "helpedAt": null,
    "id": 1,
    "location": null,
    "questionTypes": [
      QuestionTypeModel {
        "cid": 1,
        "color": "#000000",
        "id": 1,
        "name": "Question Type",
        "queueId": 2,
      },
    ],
    "queueId": 1,
    "status": "Queued",
    "taHelped": null,
    "taHelpedId": null,
    "text": "help us",
  },
}
`;
